bin/arrayNormalization.R

```diff
@@ -0,0 +1,374 @@
+			# method if they are both loaded (here they shouldn't be so don't
+			# call limma's explicitly).
+			print("Background correcting")
+			dataSet <- backgroundCorrect(dataSet, method="normexp", offset=50)
+			
+			if(mode == "agil1") {
+				
+				# Between-array normalization for 1-colour.
+				normData <- normalizeBetweenArrays(dataSet, method="quantile")
+			}
+
+			else if(mode == "agil2") {
+				# Within-array normalization for 2-colour.
+				# This gives the M-value and A-value for each gene on the array.
+				# M = log2(Cy5) - log2(Cy3)
+				# A = 0.5[log2(Cy5) + log2(Cy3)]
+				print("Running within-array normalization")
+				normData <- normalizeWithinArrays(dataSet, method="loess")
+			}
+
+			# average for probe duplicates -- some probes are on the array
+			# multiple times in different locations.
+			print("Averaging duplicated probes")
+			normData <- avereps(normData, ID=normData$genes$ProbeName)
+			
+			# Write results to file
+			if(mode == "agil1") {
+				# For single channel Agilent data, this is just the
+				# quantile-normalized expressions (analogous to RMA-normalized
+				# expressions from Affymetrix data).
+				
+				# Make a dataframe with probe IDs and normalized expressions.
+				normExprs <- data.frame(cbind(normData$genes$ProbeName, normData$E))
+				# Add assayNames as column headings
+				colnames(normExprs) <- c("DesignElementAccession", assayNames)
+
+				print("Writing normalized expressions matrix")
+				write.table(normExprs, file=outFile, sep="\t", quote=FALSE, row.names=FALSE)
+				
+				# Now files are written, function returns and script exits.
+				return("Agilent 1-colour normalization finished")
+			}
+			else if(mode == "agil2") {
+				# For two channel Agilent data, this is logFCs (M-values) and
+				# average intensities (A-values).
+
+				# data frames for M-values and A-values. Put rownames (design
+				# element IDs) as the first column and name it
+				# "DesignElementAccession".
+				mValues <- data.frame(normData$M)
+				mValues <- data.frame(cbind(rownames(mValues), mValues))
+				colnames(mValues) <- c("DesignElementAccession", assayNames)
+				aValues <- data.frame(normData$A)
+				aValues <- data.frame(cbind(rownames(aValues), aValues))
+				colnames(aValues) <- c("DesignElementAccession", assayNames)
+
+				# Filename for A-values
+				outFile_A = paste(outFile, ".A-values", sep="")
+
+				print("Writing M-values and A-values")
+				
+				# write M-values and A-values, without column names because we already wrote them above.
+				write.table(mValues, file=outFile, sep="\t", quote=FALSE, row.names=FALSE)
+				write.table(aValues, file=outFile_A, sep="\t", quote=FALSE, row.names=FALSE)
+		
+				# Now files are written, function returns and script exits.
+				return("Agilent 2-colour normalization complete")
+			}
+		}
+		# Illumina data 
+		# use lumi package for pre-processing of raw text data.
+		else if(mode == "lumi") {
```

Comment:
This seems to be like a very long function, might be better to have smaller functions for each technology, and then you all them accordingly upstream.

Reply:
yes, but it makes no sense to have a single function for three different technologies, if you had 10, it would be much longer, and every time you would need to add one, you risk breaking the others. It would be simpler to read if we do one function per technology. Breaking it up makes it more amenable for testing. Please split.